R/pseudocolor_plot.R

```diff
@@ -0,0 +1,346 @@
+#'                                     sample_type = rep(c("Unstimulated", "Stimulated"),
+#'                                                       each = 500, times = 2),
+#'                                     row.names = paste0("Cell", 1:2000)))
+#'
+#' plot_gate <-
+#'   data.frame(xmin = c(70, 75),
+#'              xmax = c(150, 155),
+#'              ymin = c(50, 50),
+#'              ymax = c(150, 150),
+#'              sample = c("A", "B"))
+#'
+#'
+#' pseudocolor_plot(object,
+#'                  marker1 = "Feature1",
+#'                  marker2 = "Feature2",
+#'                  facet_vars = "sample",
+#'                  plot_gate = plot_gate,
+#'                  layer = "counts")
+#'
+#' @export
+#'
+pseudocolor_plot <- function(
+    object,
+    marker1,
+    marker2,
+    facet_vars = NULL,
+    plot_gate = NULL,
+    scale_density = TRUE,
+    margin_density = FALSE,
+    coord.fixed = T,
+    pt.size = 1,
+    alpha = 1,
+    layer = "data",
+    grid_n = 500,
+    ...
+) {
+
+  # Validate input parameters
+  stopifnot(
+    "'facet_vars' must either be NULL or be a character vector with 1 or 2 elements" =
+      is.null(facet_vars) | length(facet_vars) %in% 1:2,
+    "Variables in 'facet_vars' must be available in the object" =
+      is.null(facet_vars) | all(facet_vars %in% colnames(object[[]])),
+
+    "'marker1' must be a character of length 1" =
+      is.character(marker1),
+    "'marker2' must be a character of length 1" =
+      is.character(marker2),
+    "'marker1' must be available in the object" =
+      marker1 %in% rownames(object),
+    "'marker2' must be available in the object" =
+      marker2 %in% rownames(object),
+
+    "'grid_n' must be a numeric of length 1" =
+      is.numeric(grid_n),
+    "'scale_density' must be either TRUE or FALSE" =
+      is.logical(scale_density),
+    "'margin_density' must be either TRUE or FALSE" =
+      is.logical(margin_density),
+
+    "'pt.size' must be a numeric of length 1" =
+      is.numeric(pt.size),
+    "'alpha' must be a numeric of length 1" =
+      is.numeric(alpha),
+    "'layer' must be a character of length 1" =
+      is.character(layer),
+    "'coord.fixed' must be either TRUE or FALSE" =
+      is.logical(coord.fixed),
+
+    "'plot_gate' must be either NULL or a data.frame" =
+      is.null(plot_gate) | is.data.frame(plot_gate),
+    "'plot_gate' must have columns 'xmin', 'xmax', 'ymin', 'ymax'" =
+      is.null(plot_gate) | all(c("xmin", "xmax", "ymin", "ymax") %in% colnames(plot_gate)),
+
+    "'object' must be a Seurat object" =
+      inherits(object, "Seurat"),
+
+    "Marginal density is not supported when 'facet_vars' is not NULL" =
+      is.null(facet_vars) | !isTRUE(margin_density)
+  )
+
+  if(isTRUE(coord.fixed) && isTRUE(margin_density)) {
+    warning("Fixed coordinates ('coord.fixed' = TRUE) is not supported when 'margin_density' is TRUE")
+  }
+
+  # Get data
+  plot_data <-
```

Comment:
I will give it a try tomorrow as well but have you tried this on a merged Seurat object? If the data contains multiple layers (as a result of merging CellGraphAssay5 objects for example), this might give some unexpected results. I had it with the Normalize function as well.
I guess since you by default use the "data" layer it should be fine, but Ludvig added a JoinLayers function that could be of help if it turns out to give a warning...

Reply:
Just tried it and no issues on a merged object, so all good!